sys/src/state.rs

```diff
@@ -81,572 +76,81 @@ pub struct caml_domain_state {
+#[repr(C)]
+#[derive(Debug)]
+#[cfg(ocaml5)]
+pub struct caml_domain_state {
+    pub _young_limit: core::sync::atomic::AtomicUsize,
+    pub _young_ptr: *mut Value,
+    pub _exception_pointer: *mut Char,
+    pub _young_start: *mut Value,
+    pub _young_end: *mut Value,
+    pub _current_stack: *mut core::ffi::c_void, // TODO: add `struct stack_info`
+    pub _exn_handler: *mut core::ffi::c_void,
+    pub _action_pending: core::ffi::c_int,
+    pub _c_stack: *mut core::ffi::c_void, // TODO: add `struct c_stack_link`
+    pub _stack_cache: *mut *mut core::ffi::c_void,
+    pub _gc_regs_buckets: *mut Value,
+    pub _gc_regs: *mut Value,
+    pub _minor_tables: *mut core::ffi::c_void, // TODO: add `struct caml_minor_tables`
+    pub _mark_stack: *mut core::ffi::c_void,   // TODO: add `struct mark_stack`
+    pub _marking_done: crate::Uintnat,
+    pub _sweeping_done: crate::Uintnat,
+    pub _allocated_words: crate::Uintnat,
+    pub _swept_words: crate::Uintnat,
+    pub _major_work_computed: crate::Intnat,
+    pub _major_work_todo: crate::Intnat,
+    pub _major_gc_clock: f64,
+    pub _local_roots: *mut crate::memory::CamlRootsBlock,
+    pub _ephe_info: *mut core::ffi::c_void,
+    pub _final_info: *mut core::ffi::c_void,
+    pub _backtrace_pos: crate::Intnat,
+    pub _backtrace_active: crate::Intnat,
+    pub _backtrace_buffer: *mut backtrace_slot,
+    pub _backtrace_last_exn: Value,
+    pub _compare_unordered: crate::Intnat,
+    pub _oo_next_id_local: crate::Uintnat,
+    pub _requested_major_slice: crate::Uintnat,
+    pub _requested_minor_slice: crate::Uintnat,
+    pub _requested_minor_gc: crate::Uintnat,
+    pub _requested_external_interrupt: core::sync::atomic::AtomicUsize,
+    pub _parser_trace: core::ffi::c_int,
+    pub _minor_heap_wsz: usize,
+    pub _shared_heap: *mut core::ffi::c_void,
+    pub _id: core::ffi::c_int,
+    pub _unique_id: core::ffi::c_int,
+    pub _dls_root: Value,
+    pub _extra_heap_resources: f64,
+    pub _extra_heap_resources_minor: f64,
+    pub _dependent_size: crate::Uintnat,
+    pub _dependent_allocated: crate::Uintnat,
+    pub _caml_extern_state: *mut core::ffi::c_void,
+    pub _caml_intern_state: *mut core::ffi::c_void,
+    pub _stat_minor_words: crate::Uintnat,
+    pub _stat_promoted_words: crate::Uintnat,
+    pub _stat_major_words: crate::Uintnat,
+    pub _stat_minor_collections: crate::Intnat,
+    pub _stat_forced_major_collections: crate::Intnat,
+    pub _stat_blocks_marked: crate::Uintnat,
+    pub _inside_stw_handler: core::ffi::c_int,
+    pub _trap_sp_off: crate::Intnat,
+    pub _trap_barrier_off: crate::Intnat,
+    pub _trap_barrier_block: i64,
+    pub _external_raise: *mut core::ffi::c_void,
+    pub _extra_params_area: [u8; 0],
 }
 
-#[cfg(not(caml_state))]
 extern "C" {
-
     #[doc(hidden)]
-    pub static mut caml_local_roots: *mut crate::memory::CamlRootsBlock;
+    pub static mut Caml_state: *mut caml_domain_state;
```

Comment:
In the case of `ocaml-sys`, not all users are required to use to boxroots library so leaving this as an option seems best. It might make sense to put this behind a feature, but I'm not convinced that's necessary.